File: prbf.py
#!env python3

import ipdb
import sys
import petsc4py
petsc4py.init(sys.argv)
from petsc4py import PETSc
from mpi4py import MPI
import numpy as np
from phelper import *

eMesh = {1: [np.linspace(0, 1, 4) ],
         2: [np.linspace(0, 0.5, 5, False), # False == Do no include endpoint in range
             np.linspace(0.5, 1, 8 )],
         4: [np.linspace(0, 0.25, 5, False),
             np.linspace(0.25, 0.5, 5, False),
             np.linspace(0.5, 0.75, 5, False),
             np.linspace(0.75, 1, 5, False), ]
         }


MPIrank = MPI.COMM_WORLD.Get_rank()
MPIsize = MPI.COMM_WORLD.Get_size()

nSupport = 5        # Number of support points
supportSpace = (-0.1, 1.1)  # Range in which the support points are equally distributed

# Dimension of interpolation. Used for adding a polynomial to the matrix. Set to zero to deactivate polynomial
dimension = 1
polyparams = dimension+1 if dimension else 0

def main():
    shuffle_mesh(eMesh)
    ePoints = eMesh[MPIsize][MPIrank] # np.array of positions to evaluate
    supports = np.linspace(supportSpace[0], supportSpace[1], nSupport)
    sPoints = partitions(supports)[MPIrank]

    A = PETSc.Mat(); A.create()
    E = PETSc.Mat(); E.create()
    if MPIrank == MPIsize-1 and dimension > 0: # The last rank gets the polynomial rows
        A.setSizes( size = ((len(sPoints)+polyparams, PETSc.DETERMINE), (len(sPoints)+polyparams, PETSc.DETERMINE)) )
        E.setSizes( size = ((len(ePoints), PETSc.DETERMINE), (len(sPoints)+polyparams, PETSc.DETERMINE)) )
    else:
        A.setSizes( size = ((len(sPoints), PETSc.DETERMINE), (len(sPoints), PETSc.DETERMINE)) )
        E.setSizes( size = ((len(ePoints), PETSc.DETERMINE), (len(sPoints), PETSc.DETERMINE)) )
    A.setName("System Matrix");  A.setFromOptions(); A.setUp()
    E.setName("Evaluation Matrix");  E.setFromOptions(); E.setUp()
    
    c = A.createVecRight(); c.setName("Coefficients")
    b = A.createVecRight(); b.setName("RHS Function Values")
    interp = E.createVecLeft(); interp.setName("interp")

    for row in range(*A.owner_range): # Rows are partioned
        if row >= len(supports): break # We are not setting the rows for the polynomial, this is done when setting each column.
        for col in range(nSupport):
            v = basisfunction(abs(supports[row]-supports[col]))
            if v != 0 or row == col: # Set 0 explicitly only on the main diagonal, petsc requirement
                A.setValue(row, col, v)
        b.setValue(row, testfunction(supports[row])) # Add the solution to the RHS

        # Add the polynomial
        if dimension:
            A.setValue(row, nSupport, 1) # Const part of the polynom
            A.setValue(nSupport, row, 1) # Ensure symmetricity
            for d in range(dimension):
                A.setValue(row, nSupport + 1 + d, supports[row]) # Value of support point
                A.setValue(nSupport + 1 + d, row, supports[row])

        
    A.assemble()
    b.assemble()
    A.view()
    A.view(PETSc.Viewer.DRAW().createDraw()) # Use command line -draw_pause <sec>.

    Print("polyparams= ", polyparams)
    Print("A Size =", A.getSize())
    Print("E Global Size = ", E.getSize())
    Print("E Local  Size = ", E.getLocalSize())
    Print("E Owner Range", E.owner_range)

    offset = E.owner_range[0]
    for row in range(*E.owner_range):
        for col in range(E.getSize()[1]-polyparams):
            E.setValue(row, col, basisfunction(abs(ePoints[row-offset] - supports[col])))
        
        # Add the polynomial
        if dimension:
            E.setValue(row, nSupport, 1)
            for d in range(dimension):
                E.setValue(row, nSupport + 1 + d, ePoints[row-offset])

    E.assemble()
    E.view(PETSc.Viewer.DRAW().createDraw()) # Use command line -draw_pause <sec>.

    ksp = PETSc.KSP()
    ksp.create()
    ksp.setOperators(A)
    ksp.setFromOptions()
    ksp.solve(b, c)

    E.mult(c, interp);

    scatter, interp0 = PETSc.Scatter.toZero(interp)
    scatter.scatter(interp, interp0)
    scatter, c0 = PETSc.Scatter.toZero(c)
    scatter.scatter(c, c0)
    
    if MPIrank == 0:
        plot(supports, eMesh, interp0.array, c0.array, dimension)
        


if __name__ == '__main__':
    main()

File: prbf.cpp
#include <iostream>
#include <map>

#include "petscksp.h"
#include "Petsc.hpp"
#include "chelper.hpp"
#include "prettyprint.hpp"

using std::cout;
using std::endl;

const int dimension = 1;
const int polyparams = dimension ? dimension+1 : 0;

const int nSupport = 10;
auto supports = linspace(-0.1, 1.1, nSupport);

std::map<int, std::vector<std::vector<double>>> eMesh = {
  {1, {linspace(0.0, 1.0, 4)} },
  {2, {linspace(0.0, 0.5, 5, false),
       linspace(0.5, 1.0, 8)} },
  {4, {linspace(0.0, 0.25, 5.0, false),
       linspace(0.25, 0.5, 5.0, false),
       linspace(0.5, 0.75, 5.0, false),
       linspace(0.75, 1.0, 5.0, false)} }
};


void RBF()
{
  int MPIrank, MPIsize;
  MPI_Comm_rank(PETSC_COMM_WORLD, &MPIrank);
  MPI_Comm_size(PETSC_COMM_WORLD, &MPIsize);

  auto sPoints = partition(supports, MPIsize)[MPIrank];
  auto ePoints = eMesh[MPIsize][MPIrank];

  petsc::Matrix A("System Matrix"), E("Evaluation Matrix");
  if ( (MPIrank == MPIsize-1) and (dimension > 0) ) {
    A.init( sPoints.size()+polyparams, sPoints.size()+polyparams, PETSC_DECIDE, PETSC_DECIDE );
    E.init( ePoints.size(), sPoints.size()+polyparams, PETSC_DECIDE, PETSC_DECIDE );
  }
  else {
    A.init( sPoints.size(), sPoints.size(), PETSC_DECIDE, PETSC_DECIDE );
    E.init( ePoints.size(), sPoints.size(), PETSC_DECIDE, PETSC_DECIDE );
  }
  petsc::Vector c(A, "Coefficients", petsc::Vector::RIGHT);
  petsc::Vector b(A, "RHS Function Values", petsc::Vector::RIGHT);
  petsc::Vector interp(E, "Interpolation Results", petsc::Vector::LEFT);

  PetscInt rangeStart, rangeEnd;
  std::tie(rangeStart, rangeEnd) = A.ownerRange();

  cout << "[" << MPIrank << "] Original Range Start / End = " << rangeStart << ", " << rangeEnd << endl;

  // Introduce some overlap for testing.
  rangeStart -= 2;
  rangeEnd += 2;
  if (rangeStart < 0) rangeStart = 0;
  if (rangeEnd > nSupport+polyparams) rangeEnd = nSupport+polyparams;
  
  cout << "[" << MPIrank << "] Range Start / End = " << rangeStart << ", " << rangeEnd << endl;
    
  for (PetscInt row = rangeStart; row < rangeEnd; row++) {
    if (row >= supports.size())
      break;
    for (PetscInt col = 0; col < nSupport; col++) {
      auto v = basisfunction( std::abs(supports[row]-supports[col]));
      if (v != 0)
        A.setValue(row, col, v);
    }
    b.setValue(row, testfunction(supports[row]));

    if (dimension) { // add the polynomial
      A.setValue(row, nSupport, 1);
      A.setValue(nSupport, row, 1); // Ensure symmetricity, might not be needed when sbaij is used
      for (PetscInt d = 0; d < dimension; d++ ) {
        A.setValue(row, nSupport + 1 + d, supports[row]);
        A.setValue(nSupport + 1 + d, row, supports[row]);
      }
    }
  }
  A.assemble(MAT_FLUSH_ASSEMBLY);
  petsc::Vector zeros(A); // Petsc requires that all diagonal entries are set, even if set to zero.
  MatDiagonalSet(A.matrix, zeros.vector, ADD_VALUES);
  A.assemble();
  b.assemble();
  // A.view();
  // A.viewDraw();  

  std::tie(rangeStart, rangeEnd) = E.ownerRange();
  for (PetscInt row = rangeStart; row < rangeEnd; row++) {
    for (PetscInt col = 0; col < E.getSize().second; col++)
      E.setValue(row, col, basisfunction( std::abs(ePoints[row-rangeStart]-supports[col])) );

    if (dimension) { // add the polynomail
      E.setValue(row, nSupport, 1);
      for (PetscInt d = 0; d < dimension; d++) {
        E.setValue(row, nSupport + 1 + d, ePoints[row-rangeStart]);
      }
    }
  }
  E.assemble();
  // E.view();
  // E.viewDraw();
  // b.view(); 
  KSP ksp;
  KSPCreate(PETSC_COMM_WORLD, &ksp);
  KSPSetOperators(ksp, A.matrix, A.matrix);
  KSPSetFromOptions(ksp);

  KSPSolve(ksp, b.vector, c.vector);

  MatMult(E.matrix, c.vector, interp.vector);
  c.view();
  interp.view();
}




int main(int argc, char *argv[])
{
  PetscInitialize(&argc, &argv, NULL, NULL);
  RBF();  
  PetscFinalize();  
  return 0;
}
